File: src/main/java/com/paragon/client/ui/configuration/panel/element/setting/SliderElement.java
package com.paragon.client.ui.configuration.panel.element.setting;

import com.paragon.api.setting.Setting;
import com.paragon.api.util.calculations.MathsUtil;
import com.paragon.api.util.render.RenderUtil;
import com.paragon.api.setting.Bind;
import com.paragon.client.ui.util.animation.Animation;
import com.paragon.client.ui.util.animation.Easing;
import com.paragon.client.ui.util.Click;
import com.paragon.client.ui.configuration.panel.element.Element;
import com.paragon.client.ui.configuration.panel.element.module.ModuleElement;
import net.minecraft.util.math.MathHelper;
import org.lwjgl.input.Mouse;

import java.awt.Color;

@SuppressWarnings("unchecked")
public final class SliderElement extends Element {

    private final Setting<Number> setting;

    private boolean dragging = false;

    private final Animation scrollAnimation = new Animation(() -> 1250f, false, () -> Easing.LINEAR);

    public SliderElement(int layer, Setting<Number> setting, ModuleElement moduleElement, float x, float y, float width, float height) {
        super(layer, x, y, width, height);

        setParent(moduleElement.getParent());
        this.setting = setting;

        setting.getSubsettings().forEach(subsetting -> {
            if (subsetting.getValue() instanceof Boolean) {
                getSubElements().add(new BooleanElement(layer + 1, (Setting<Boolean>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof Enum<?>) {
                getSubElements().add(new EnumElement(layer + 1, (Setting<Enum<?>>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof Number) {
                getSubElements().add(new SliderElement(layer + 1, (Setting<Number>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof Bind) {
                getSubElements().add(new BindElement(layer + 1, (Setting<Bind>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof Color) {
                getSubElements().add(new ColourElement(layer + 1, (Setting<Color>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }

            else if (subsetting.getValue() instanceof String) {
                getSubElements().add(new StringElement(layer + 1, (Setting<String>) subsetting, moduleElement, getX(), getY(), getWidth(), getHeight()));
            }
        });
    }

    @Override
    public void render(int mouseX, int mouseY, int dWheel) {
        if (setting.isVisible()) {
            float renderWidth = 0;
            float maxWidth = getWidth() - getLayer() * 2;

            if (setting.getValue() instanceof Float) {
                // Set values
                float diff = Math.min(maxWidth, Math.max(0, mouseX - (getX() + getLayer())));

                float min = setting.getMin().floatValue();
                float max = setting.getMax().floatValue();

                renderWidth = maxWidth * (setting.getValue().floatValue() - min) / (max - min);

                if (!Mouse.isButtonDown(0)) {
                    dragging = false;
                }

                if (dragging) {
                    if (diff == 0) {
                        setting.setValue(setting.getMin());
                    } else {
                        float newValue = (float) MathsUtil.roundDouble(((diff / maxWidth) * (max - min) + min), 2);

                        float precision = 1 / setting.getIncrementation().floatValue();
                        newValue = Math.round(Math.max(min, Math.min(max, newValue)) * precision) / precision;

                        setting.setValue(newValue);
                    }
                }
            } else if (setting.getValue() instanceof Double) {
                // Set values
                double diff = Math.min(maxWidth, Math.max(0, mouseX - (getX() + getLayer())));

                double min = setting.getMin().doubleValue();
                double max = setting.getMax().doubleValue();

                renderWidth = (float) (maxWidth * (setting.getValue().doubleValue() - min) / (max - min));

                if (!Mouse.isButtonDown(0)) {
                    dragging = false;
                }

                if (dragging) {
                    if (diff == 0) {
                        setting.setValue(setting.getMin());
                    } else {
                        double newValue = MathsUtil.roundDouble(((diff / maxWidth) * (max - min) + min), 2);

                        double precision = 1 / setting.getIncrementation().floatValue();
                        newValue = Math.round(Math.max(min, Math.min(max, newValue)) * precision) / precision;

                        setting.setValue(newValue);
                    }
                }
            }

            RenderUtil.drawRect(getX(), getY(), getWidth(), getHeight(), new Color(40, 40, 45).getRGB());
            RenderUtil.drawRect(getX() + getLayer(), getY(), getWidth() - getLayer() * 2, getHeight(), new Color((int) (40 + (30 * getHover().getAnimationFactor())), (int) (40 + (30 * getHover().getAnimationFactor())), (int) (45 + (30 * getHover().getAnimationFactor()))).getRGB());

            RenderUtil.drawRect(getX() + getLayer(), getY(), renderWidth, getHeight(), Color.HSBtoRGB(getParent().getLeftHue() / 360, 1f, (float) (0.5f + (0.25f * getHover().getAnimationFactor()))));

            float x = getX() + (getLayer() * 2) + 5;
            float totalWidth = getWidth() - (getLayer() * 2);
            float maxTextWidth = totalWidth - getStringWidth(setting.getValue().toString()) - 5;

            float visibleX = getStringWidth(setting.getName()) - maxTextWidth;

            scrollAnimation.setState(isHovered(mouseX, mouseY));

            if (getStringWidth(setting.getName()) > maxTextWidth) {
                x -= (visibleX + 9) * scrollAnimation.getAnimationFactor();
            }

            float scissorY = MathHelper.clamp(getY(), getParent().getY() + 22, getParent().getY() + MathHelper.clamp(
                    // Scissor comedy
                    getParent().getScissorHeight() + 8, 0, 358));

            float scissorHeight = getHeight();

            RenderUtil.startGlScissor(getX() + (getLayer() * 2), scissorY, totalWidth - (getStringWidth(setting.getValue().toString()) + 9), scissorHeight);

            renderText(setting.getName(), x, getY() + getHeight() / 2 - 3.5f, 0xFFFFFFFF);

            RenderUtil.endGlScissor();

            renderText(setting.getValue().toString(), getX() + getWidth() - (getLayer() * 2) - getStringWidth(setting.getValue().toString()) - 3, getY() + getHeight() / 2 - 3.5f, 0xFFFFFFFF);

            super.render(mouseX, mouseY, dWheel);
        }
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, Click click) {
        if (setting.isVisible()) {
            if (isHovered(mouseX, mouseY) && getParent().isElementVisible(this) && click.equals(Click.LEFT)) {
                this.dragging = true;
            }

            super.mouseClicked(mouseX, mouseY, click);
        }
    }

    @Override
    public void mouseReleased(int mouseX, int mouseY, Click click) {
        dragging = false;

        if (setting.isVisible()) {
            super.mouseReleased(mouseX, mouseY, click);
        }
    }

    @Override
    public void keyTyped(int keyCode, char keyChar) {
        if (setting.isVisible()) {
            super.keyTyped(keyCode, keyChar);
        }
    }

    @Override
    public float getHeight() {
        return getSetting().isVisible() ? super.getHeight() : 0;
    }

    @Override
    public float getSubElementsHeight() {
        return getSetting().isVisible() ? super.getSubElementsHeight() : 0;
    }

    @Override
    public float getTotalHeight() {
        return getSetting().isVisible() ? super.getTotalHeight() : 0;
    }

    public boolean isDragging() {
        return dragging;
    }

    public Setting<Number> getSetting() {
        return setting;
    }

}

File: src/main/kotlin/com/paragon/client/ui/configuration/zeroday/element/setting/SliderElement.kt
package com.paragon.client.ui.configuration.zeroday.element.setting

import com.paragon.Paragon
import com.paragon.api.event.client.SettingUpdateEvent
import com.paragon.api.setting.Bind
import com.paragon.api.setting.Setting
import com.paragon.api.util.calculations.MathsUtil.roundDouble
import com.paragon.api.util.render.RenderUtil.drawRect
import com.paragon.api.util.render.RenderUtil.popScissor
import com.paragon.api.util.render.RenderUtil.pushScissor
import com.paragon.api.util.render.font.FontUtil.drawStringWithShadow
import com.paragon.api.util.render.font.FontUtil.getStringWidth
import com.paragon.client.ui.configuration.zeroday.element.Element
import com.paragon.client.ui.configuration.zeroday.element.module.ModuleElement
import com.paragon.client.ui.util.Click
import me.surge.animation.Animation
import me.surge.animation.Easing
import net.minecraft.util.math.MathHelper
import org.lwjgl.input.Mouse
import java.awt.Color
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

class SliderElement(
    layer: Int,
    setting: Setting<Number>,
    moduleElement: ModuleElement,
    x: Float,
    y: Float,
    width: Float,
    height: Float
) : Element(layer, x, y, width, height) {

    override var height: Float
        get() = if (setting.isVisible()) super.height else 0f
        set(value) {}

    val setting: Setting<Number>
    var isDragging = false
        private set

    private val scrollAnimation = Animation({ 1250f }, false) { Easing.LINEAR }

    init {
        parent = moduleElement.parent
        this.setting = setting
        setting.subsettings.forEach {
            when (it.value) {
                is Boolean -> subElements.add(
                    BooleanElement(
                        layer + 1,
                        (it as Setting<Boolean>),
                        moduleElement,
                        x,
                        y,
                        width,
                        height
                    )
                )

                is Enum<*> -> subElements.add(
                    EnumElement(
                        layer + 1,
                        (it as Setting<Enum<*>>),
                        moduleElement,
                        x,
                        y,
                        width,
                        height
                    )
                )

                is Number -> subElements.add(
                    SliderElement(
                        layer + 1,
                        it as Setting<Number>,
                        moduleElement,
                        x,
                        y,
                        width,
                        height
                    )
                )

                is Bind -> subElements.add(
                    BindElement(
                        layer + 1,
                        (it as Setting<Bind>),
                        moduleElement,
                        x,
                        y,
                        width,
                        height
                    )
                )

                is Color -> subElements.add(
                    ColourElement(
                        layer + 1,
                        it as Setting<Color>,
                        moduleElement,
                        x,
                        y,
                        width,
                        height
                    )
                )

                is String -> subElements.add(
                    StringElement(
                        layer + 1,
                        it as Setting<String>,
                        moduleElement,
                        x,
                        y,
                        width,
                        height
                    )
                )
            }
        }
    }

    override fun render(mouseX: Int, mouseY: Int, dWheel: Int) {
        if (setting.isVisible()) {
            var renderWidth = 0f
            val maxWidth = width - layer * 2
            if (setting.value is Float) {
                // Set values
                val diff = min(maxWidth, max(0f, mouseX - (x + layer)))

                val min = setting.min.toFloat()
                val max = setting.max.toFloat()

                renderWidth = maxWidth * ((setting.value as Float) - min) / (max - min)

                if (!Mouse.isButtonDown(0)) {
                    isDragging = false
                }
                if (isDragging) {
                    if (diff == 0f) {
                        setting.setValue(setting.min)
                        Paragon.INSTANCE.eventBus.post(SettingUpdateEvent(setting))
                    } else {
                        var newValue = roundDouble((diff / maxWidth * (max - min) + min).toDouble(), 2).toFloat()
                        val precision = 1 / setting.incrementation.toFloat()

                        newValue = (max(min, min(max, newValue)) * precision).roundToInt() / precision

                        setting.setValue(newValue)
                        Paragon.INSTANCE.eventBus.post(SettingUpdateEvent(setting))
                    }
                }
            } else if (setting.value is Double) {
                // Set values
                val diff = min(maxWidth, max(0f, mouseX - (x + layer))).toDouble()

                val min = setting.min.toDouble()
                val max = setting.max.toDouble()

                renderWidth = (maxWidth * ((setting.value as Double) - min) / (max - min)).toFloat()

                if (!Mouse.isButtonDown(0)) {
                    isDragging = false
                }
                if (isDragging) {
                    if (diff == 0.0) {
                        setting.setValue(setting.min)
                        Paragon.INSTANCE.eventBus.post(SettingUpdateEvent(setting))
                    } else {
                        var newValue = roundDouble(diff / maxWidth * (max - min) + min, 2)
                        val precision = (1 / setting.incrementation.toFloat()).toDouble()
                        newValue = (max(min, min(max, newValue)) * precision).roundToInt() / precision
                        setting.setValue(newValue)
                        Paragon.INSTANCE.eventBus.post(SettingUpdateEvent(setting))
                    }
                }
            }

            drawRect(x, y, width, height, Color(40, 40, 45).rgb)

            drawRect(
                x + layer,
                y,
                width - layer * 2,
                height,
                Color(
                    (40 + 30 * hover.getAnimationFactor()).toInt(),
                    (40 + 30 * hover.getAnimationFactor()).toInt(),
                    (45 + 30 * hover.getAnimationFactor()).toInt()
                ).rgb
            )

            drawRect(
                x + layer,
                y,
                renderWidth,
                height,
                Color.HSBtoRGB(parent.leftHue / 360, 1f, (0.5f + 0.25f * hover.getAnimationFactor()).toFloat())
            )

            val x = x + layer * 2 + 5
            val totalWidth = width - layer * 2
            val maxTextWidth = totalWidth - getStringWidth(setting.value.toString()) - 12
            val visibleX = getStringWidth(setting.name) - maxTextWidth
            scrollAnimation.state = isHovered(mouseX, mouseY)

            pushScissor(x.toDouble(), MathHelper.clamp(y.toDouble(), parent.y + parent.height.toDouble(), 100000.0), maxTextWidth.toDouble(), ((parent.y + parent.height) + parent.scissorHeight) - y.toDouble())

            drawStringWithShadow(setting.name, x - if (getStringWidth(setting.name) > maxTextWidth) ((visibleX + 9) * scrollAnimation.getAnimationFactor()).toFloat() else 0f, y + height / 2 - 3.5f, -0x1)

            popScissor()

            drawStringWithShadow(
                setting.value.toString(),
                (x + totalWidth) - getStringWidth(setting.value.toString()) - 10,
                y + height / 2 - 3.5f,
                -0x1
            )

            super.render(mouseX, mouseY, dWheel)
        }
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, click: Click) {
        if (setting.isVisible()) {
            if (isHovered(mouseX, mouseY) && parent.isElementVisible(this) && click == Click.LEFT) {
                isDragging = true
            }
            super.mouseClicked(mouseX, mouseY, click)
        }
    }

    override fun mouseReleased(mouseX: Int, mouseY: Int, click: Click) {
        isDragging = false
        if (setting.isVisible()) {
            super.mouseReleased(mouseX, mouseY, click)
        }
    }

    override fun keyTyped(keyCode: Int, keyChar: Char) {
        if (setting.isVisible()) {
            super.keyTyped(keyCode, keyChar)
        }
    }

    override fun getSubElementsHeight(): Float {
        return if (setting.isVisible()) super.getSubElementsHeight() else 0f
    }

    override fun getTotalHeight(): Float {
        return if (setting.isVisible()) super.getTotalHeight() else 0f
    }
}